components/brave_federated/task/model.cc

```diff
@@ -0,0 +1,275 @@
+        LinearAlgebraUtil::TransposeMatrix(x), err);
+    d_w = LinearAlgebraUtil::MultiplyVectorScalar(d_w, (-2.0 / batch_size_));
+
+    const float d_b =
+        (-2.0 / batch_size_) * std::accumulate(err.begin(), err.end(), 0.0);
+
+    weights_ = LinearAlgebraUtil::SubtractVector(
+        p_w, LinearAlgebraUtil::MultiplyVectorScalar(d_w, learning_rate_));
+    bias_ = p_b - learning_rate_ * d_b;
+
+    if (iteration % 250 == 0) {
+      auto loss_pred = Predict(x);
+      if (!loss_pred.has_value()) {
+        std::string error_message =
+            "FL: Train loss predict failed in iteration " +
+            std::to_string(iteration) + " of " +
+            std::to_string(num_iterations_);
+        return base::unexpected(error_message);
+      }
+      training_loss = ComputeNegativeLogLikelihood(y, loss_pred.value());
+    }
+
+    data_prep_duration +=
+        base::TimeDelta(execution_start_thread_ticks - data_start_thread_ticks);
+    training_duration += base::TimeDelta(GetThreadTicksIfSupported() -
+                                         execution_start_thread_ticks);
+  }
+
+  float accuracy = training_loss;
+
+  std::map<std::string, double> metrics = {
+      {"data_prep_duration_in_seconds", data_prep_duration.InSecondsF()},
+      {"training_duration_in_seconds", training_duration.InSecondsF()}};
+
+  std::vector<Weights> reported_model = {weights_, {bias_}};
+  return PerformanceReport(train_dataset.size(),  // dataset_size
+                           training_loss,         // loss
+                           accuracy,              // accuracy
+                           reported_model,        // parameters
+                           metrics                // metrics
+  );
+}
+
+base::expected<PerformanceReport, std::string> Model::Evaluate(
+    const DataSet& test_dataset) {
+  if (test_dataset.empty()) {
+    return base::unexpected("FL: Test data empty");
+  }
+
+  auto data_start_ts = GetThreadTicksIfSupported();
+
+  int num_features = test_dataset.at(0).size();
+  DataSet features(test_dataset.size(), std::vector<float>(num_features));
+  std::vector<float> ground_truth(test_dataset.size());
+
+  for (size_t i = 0; i < test_dataset.size(); i++) {
+    std::vector<float> point = test_dataset.at(i);
+    ground_truth.at(i) = point.back();
+    point.pop_back();
+    features.at(i) = point;
+  }
+
+  auto exec_start_ts = GetThreadTicksIfSupported();
+
+  auto predicted = Predict(features);
+  if (!predicted.has_value()) {
+    return base::unexpected("FL: Evaluate predict failed");
+  }
+
+  std::vector<float> predicted_value = predicted.value();
+  int total_correct = 0;
+  for (size_t i = 0; i < test_dataset.size(); i++) {
+    if (predicted_value.at(i) >= threshold_) {
+      predicted_value.at(i) = 1.0;
+    } else {
+      predicted_value.at(i) = 0.0;
+    }
+
+    if (predicted_value.at(i) == ground_truth.at(i)) {
+      total_correct++;
+    }
+  }
+  float accuracy = total_correct / test_dataset.size();
```

Comment:
Bug prone integer division

Reply:
wrapped it in a `static_cast`, is that okay?